torch/distributed/_tensor/dispatch.py

```diff
@@ -148,15 +150,111 @@ def operator_dispatch(
+                    return torch.tensor(
+                        [],
+                        dtype=dtype,
+                    )
+            else:
+                raise RuntimeError(
+                    f"{spec} has no tensor metadata."
+                )
+
+        def _default_ret_value(arg):
+            type_str = str(arg.type)
+            if type_str == "()":  # used in aten function schema for None return value
+                return None
+
+            if _arg_type_is_primitive(type_str):
+                obj_list = [None for _ in range(dist.get_world_size())]
+                dist.all_gather_object(obj_list, None)
+                obj_list = list(filter(lambda x: x is not None, obj_list))
+                # TODO: this nested if-else statement is due to MYPY's
+                # issue on inferring reduce op's type. We have to explicitly
+                # type the lambda to make MYPY happy. But we better find a
+                # smarter way to do it.
+                if type_str == "bool":
+                    return functools.reduce(lambda x, y: x and y, obj_list, True)  # type: ignore[arg-type]
+                elif (
+                    (type_str == "int")
+                    or (type_str == "float")
+                    or (type_str == "double")
+                ):
+                    return functools.reduce(lambda x, y: x + y, obj_list, 0)  # type: ignore[arg-type]
+                else:
+                    raise NotImplementedError(
+                        f"Unsupported primitive type {type_str}."
+                    )
+
+            if type_str == "Tensor":
+                assert isinstance(output_sharding.output_spec, DTensorSpec), (
+                    f"output spec does not match with output! Expected "
+                    f"DTensorSpec, got {output_sharding.output_spec}"
+                )
+                return _default_tensor(output_sharding.output_spec)
+            elif type_str == "List[Tensor]":
+                assert isinstance(output_sharding.output_spec, Sequence), (
+                    f"output spec does not match with output! Expected "
+                    f"Sequence[DTensorSpec], got {output_sharding.output_spec}"
+                )
+                return [
+                    _default_tensor(spec)
+                    for spec in output_sharding.output_spec
+                ]
+            else:
+                # TODO: add more types
+                raise NotImplementedError(
+                    f"default value for type {type_str} on non-participating"
+                    f" device is not implemented."
+                )
+
+        ret_list = op_schema.func_schema.returns
+        if len(ret_list) == 1:
+            # return: scalar/torch.Tensor
+            local_results = _default_ret_value(ret_list[0])
+        elif len(ret_list) > 1:
+            # return: List[torch.Tensor]
+            local_results = [_default_ret_value(arg) for arg in ret_list]
```

Comment:
I don't quite understand, if you already return a default tensor list in `_default_ret_value`, wouldn't this additional list comprehension become wrong?